### DISCUSSION TITLE
Simple browser in incognito mode

### DISCUSSION THREAD

**Discussion by mbowles201** [cefsharp/CefSharp#4101] [General]
I need to create browser instance within WPF that runs the chromium browser in incognito mode (cachepath = "").  I dont need any other functionality (or can get it from the examples).  I just dont see anything on setting that cachepath outside of the docs (no example).

**Comment by amaitland** [comment]
CachePath is empty by default. There's no need to do anything if that's your required behaviour.

**Reply by mbowles201** [reply]
So, each new tab in the example is in "incognito" mode from the start?

**Reply by amaitland** [reply]
No, the examples all set a `CachePath`.

**Reply by mbowles201** [reply]
Ok, I see the settings.CachePath set up in the CefExample.cs class in the CefSharp.Example.netcore project.   If I start from a blank WPF project and simply add the xaml to add a chromium browser to my project, it will be incognito by default ?

**Reply by amaitland** [reply]
> If I start from a blank WPF project and simply add the xaml to add a chromium browser to my project, it will be incognito by default ?

Yes, an in memory cache will be used by default.

**Reply by mbowles201** [reply]
Great, my goal is to open multiple instances (like the tabs in the example) and each have no cache so that Sessions do not carry over to each instance.  I want my web app to have a new session for each "tab".   Thanks!

**Reply by amaitland** [reply]
If you need to isolate each `ChromiumWebBrowser` instance then you will need to use a different `RequestContext` for each, otherwise they will by default share the same `Global Request Context`.

See https://github.com/cefsharp/CefSharp/wiki/General-Usage#request-context-browser-isolation

**Reply by mbowles201** [reply]
Thank you.

**Reply by mbowles201** [reply]
Using IWpfWebBrowser, you can not set the RequestContext.  Can you pass a ChromiumWebBrowser to the xaml ChromiumWebBrowser tag as the WebBrowser value and will that produce the affect I want?
The Request context, when added, throws this: 'Could not load file or assembly 'CefSharp.Core.Runtime, Version=101.0.150.0, Culture=neutral, PublicKeyToken=40c4b6fc221f4138'. An attempt was made to load a program with an incorrect format.'

**Reply by amaitland** [reply]
> Using IWpfWebBrowser, you can not set the RequestContext.

Correct. You need to set the `ChromiumWebBrowser.RequestContext` property.

> Can you pass a ChromiumWebBrowser to the xaml ChromiumWebBrowser tag as the WebBrowser value and will that produce the affect I want?

I have absoutely no idea what you mean by this.

If you want the same behaviour for all `ChromiumWebBrowser` instances then I'd simply subclass `ChromiumWebBrowser` and create a `CustomChromiumWebBrowser` class that encapsulates your specific requirements.

> The Request context, when added, throws this: 'Could not load file or assembly 'CefSharp.Core.Runtime, Version=101.0.150.0, Culture=neutral, PublicKeyToken=40c4b6fc221f4138'. An attempt was made to load a program with an incorrect format.'

It's a very common exception and without context/example I've got no idea what the actual case is. Design time? Runtime? What does your code look like? etc. Need more detail.

**Reply by mbowles201** [reply]
For the error:
I am using this code snippet in my BrowserTabViewModel to create a new browser per WPF tab space (like example).
This runs.  If I uncomment out the last 3 lines involving RequestContext, it builds fine, but throws that error at runtime.

`private void SetBrowser()
        {
            CefPath = Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData) +                      "\\uniquevalueperinstancehere\\Cache\\Path"; 

            if (!Directory.Exists(CefPath)) Directory.CreateDirectory(CefPath);

            var settings = new CefSettings();
            settings.CachePath = CefPath;

            if (!Cef.IsInitialized) Cef.Initialize(settings);

            Browser = new ChromiumWebBrowser();  // set above on model instantiate.

            //var requestContextSettings = new RequestContextSettings();
            //requestContextSettings.CachePath = CefPath;

            //Browser.RequestContext = new RequestContext(requestContextSettings);
        }`

**Reply by mbowles201** [reply]
ok, seems my RuntimeIdentifier had a type (though it did not kill the project), so I have it working with the code above.  Unfortunately, I still get a shared session across 2+ browser instances and I believe I am using 2 different RequestContexts.  Thoughts?

**Reply by mbowles201** [reply]
I add this to my method, but the session still persists across browser instances.

`requestContextSettings.PersistSessionCookies = false;
requestContextSettings.PersistUserPreferences = false;`

**Reply by mbowles201** [reply]
Update, I followed what this guy did (https://stackoverflow.com/questions/55707524/how-to-store-and-restore-cookies-for-multiple-browser-instances-in-cefsharp) and created 2 instances of the browser in XAML, then set contexts after the Initialize() call with different cachepaths and I get the result I am looking for.   I was following an MVVM pattern like the examples but the distinct cache and session would never work.

**Reply by amaitland** [reply]
From memory the ChromiumWebBrowser instance is always created in the view in the examples, which is what I'd recommend. Creating a UserControl or a class inheriting from ChromiumWebBrowser to encapsulate the functionality is the general approach I recommend. 

Best to avoid creating it in your viewmodel.

**Reply by mbowles201** [reply]
Thank you.  I found the MVVM approach would not work.   Now to work some WPF magic so I can allow the add and removal of tabs via code-behind.  Should be fun.

**Reply by amaitland** [reply]
If you are using the built in tabcontrol then make sure to read https://stackoverflow.com/questions/9794151/stop-tabcontrol-from-recreating-its-children

**Reply by mbowles201** [reply]
Thanks for that link.  Right now I have a couple of xaml tags that I set the requestcontext  right after Initialize().   That is working the way I want.  I dont have it bound to a list, though, I guess I could do all the requestcontexts and then put them in a list (or while putting in the list) and test that.
